subprojects/logging/src/main/java/org/gradle/internal/logging/console/LogGroupingOutputEventListener.java

```diff
@@ -0,0 +1,229 @@
+import org.gradle.internal.logging.events.BatchOutputEventListener;
+import org.gradle.internal.logging.events.CategorisedOutputEvent;
+import org.gradle.internal.logging.events.EndOutputEvent;
+import org.gradle.internal.logging.events.LogEvent;
+import org.gradle.internal.logging.events.OperationIdentifier;
+import org.gradle.internal.logging.events.OutputEvent;
+import org.gradle.internal.logging.events.ProgressCompleteEvent;
+import org.gradle.internal.logging.events.ProgressEvent;
+import org.gradle.internal.logging.events.ProgressStartEvent;
+import org.gradle.internal.logging.events.RenderableOutputEvent;
+import org.gradle.internal.logging.events.StyledTextOutputEvent;
+import org.gradle.internal.logging.text.StyledTextOutput;
+import org.gradle.internal.progress.BuildOperationType;
+import org.gradle.internal.time.TimeProvider;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+
+public class LogGroupingOutputEventListener extends BatchOutputEventListener {
+    static final int SCHEDULER_CHECK_PERIOD_MS = 5000;
+    public static final String EOL = SystemProperties.getInstance().getLineSeparator();
+    private final BatchOutputEventListener listener;
+    private final ScheduledExecutorService executor;
+    private final TimeProvider timeProvider;
+    private final Object lock = new Object();
+
+    // Allows us to map progress and complete events to start events
+    private final Map<OperationIdentifier, Object> progressIdToBuildOperationIdMap = new LinkedHashMap<OperationIdentifier, Object>();
+
+    // Maintain a hierarchy of all build operation ids, so that child operations are grouped under their ancestor
+    private final Map<Object, Object> buildOperationIdForest = new LinkedHashMap<Object, Object>();
+
+    // Event groups that are in-progress and have not been completed
+    private final Map<Object, ArrayList<CategorisedOutputEvent>> outputEventGroups = new LinkedHashMap<Object, ArrayList<CategorisedOutputEvent>>();
+    private Object lastRenderedOperationId;
+    private boolean needsHeader;
+
+    public LogGroupingOutputEventListener(BatchOutputEventListener listener, TimeProvider timeProvider) {
+        this(listener, Executors.newSingleThreadScheduledExecutor(), timeProvider);
+    }
+
+    LogGroupingOutputEventListener(BatchOutputEventListener listener, ScheduledExecutorService executor, TimeProvider timeProvider) {
+        this.listener = listener;
+        this.executor = executor;
+        this.timeProvider = timeProvider;
+    }
+
+    @Override
+    public void onOutput(OutputEvent event) {
+        synchronized (lock) {
+            if (event instanceof EndOutputEvent) {
+                renderAllGroups(outputEventGroups);
+                listener.onOutput(event);
+//                executor.shutdown();
```

Comment:
Why would we not shut down the thread pool? Isn't that the event that finalizes the build?